Container State Machine

kflansburg opened the pull request.

kflansburg commented:
Not quite ready to merge but I think it’s a good time for some reviews. What is working:

* Refactored state API to be generic.
* Rebased with master.
* Implemented WASI container state machine. This was nasty but I really like how simple the state handlers are now.

Still left to do:
* wascc container state machine.
* Some todos that I identified which probably don’t need to go into this PR.
* Remove some unwraps

thomastaylor312 reviewed (commented):
This is looking fairly solid, think we should probably get @itowlson's review on this too since he has been mucking around in here recently. I actually like the combined shared state thing as well

kflansburg commented on crates/kubelet/src/container/state.rs:
```diff
@@ -0,0 +1,106 @@
+pub async fn run_to_completion<S: ResourceState<Manifest = Container, Status = Status>>(
+    client: &kube::Client,
+    initial_state: impl State<S>,
+    shared_state: SharedState<S::SharedState>,
```
It is definitely confusing. I wonder if the `State` trait should actually be like `Handler`, and we might be able to just called it `shared` or something. Any ideas?

thomastaylor312 replied:
Because this is billed as a "state machine" (which it is), I think the name "state" should remain with the actual state type. However, just using `shared` might be just fine

kflansburg replied:
Will do.

kflansburg commented on crates/kubelet/src/lib.rs:
```diff
@@ -10,7 +10,11 @@
 //! use kubelet::pod::Pod;
 //! use kubelet::provider::Provider;
-//! use kubelet::state::{SharedState, Stub, AsyncDrop};
+//! use kubelet::state::{SharedState};
```
I was going to agree, but on second thought I think we should stick to what `rustfmt` enforces. Adding additional style opinions would be a lot of manual labor to enforce consistently. Is that ok?

thomastaylor312 replied:
I don't think `rustfmt` enforces this. I very rarely see single imports using braces and rustfmt doesn't correct me if I don't have them

kflansburg replied:
Yeah, I think my point is that we cant automatically enforce that style either way. I will remove it here but I usually think that opinions on style held by a project should be something that can be automatically linted.

kflansburg replied:
For posterity, I used this to find one other occurrence:
```
grep -rE . -e '{([a-zA-Z]+)}'
```
Something like that could be added to our test suite.

kflansburg commented on crates/kubelet/src/pod/status.rs:
```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
-    fn default() -> Self {
-        Self::LeaveUnchanged
+    /// Set Pod phase.
+    pub fn set_phase(&mut self, phase: Phase) {
```
I think we could basically rename this struct to `StatusBuilder` then. I can if you want. The reason I did this is that there is a distinction between a field that was never set, or is explicitly being set to "", this results in patches overwriting things they shouldnt. There might be a simpler way to do this, and I didnt do the same thing for containers because its much less important. Thoughts?

thomastaylor312 replied:
I personally think this:

```rust
let status = StatusBuilder::new().phase(...).message(...).build();
```

reads much better (and is equivalent to):

```rust
let mut status = Status::new();
status.set_phase(...);
status.set_message(...);
```

I figure the point about an empty field is pretty easy to handle. You can still use the setter/builder to explicitly set an empty string and leave it empty otherwise. However, this is completely a bikeshedding, so feel no pressure to change it

kflansburg replied:
Yep, I will refactor it to a Builder, I was just explaining why it got so complicated.

kflansburg commented on crates/kubelet/src/pod/status.rs:
```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
+    }
+
+    pub(crate) fn into_json(self) -> serde_json::Value {
+        let mut status = serde_json::Map::new();
```
See above. Calling `json!` doesnt really create a patch. It may be that what we really want is to use `jsondiff` here, but then you need to have an updated version of the Pod status and then there are race conditions, and this actually seems simpler.

thomastaylor312 replied:
Ah, ok valid point

kflansburg commented on crates/kubelet/src/pod/status.rs:
```diff
@@ -213,6 +239,12 @@ pub enum Phase {
 
+impl std::fmt::Display for Phase {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", serde_json::json!(self).as_str().unwrap())
```
I'm not sure what implementation you are referring to. I don't see that `PodStatus` implements `Display`, and phase is simply stored as a `String`. The reason I did it this way is that as far as I know this is the only way to automatically generate `Display` for `enums`, rather than having to specify the exact string used. You can also use `Debug`, but `serde` lets us control the capitalization, etc. a little more. I don't think this can panic. I can change this to use `Debug` but I don't like the idea of having to hard code the strings and update them if the `enum` changes.

thomastaylor312 replied:
Oh wait, I somehow misread `Phase` as `Status` when I was going through your code 🤦. Sorry

kflansburg commented on crates/kubelet/src/state/common/mod.rs:
```diff
@@ -50,21 +52,21 @@ pub trait GenericProviderState: 'static + Send + Sync {
     /// Stores the pod module binaries for future execution. Typically your
     /// implementation can just move the modules map into a member field.
-    fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
+    async fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
```
So, I ran into this because the `ModuleRunContext` had to be moved into a `SharedState` so that individual container state machines could load their module data. I'm glad you brought this up because I think that we should make pretty much all public trait methods `async`, the reason being that we have a lot of `async` methods and Tokio `RwLocks` floating around and it is really annoying as a provider author to realize that you can't use something `async` in an `async` trait just because we (the maintainers) didn't forsee that use case. Thoughts?

thomastaylor312 replied:
I figured that may be part of the reasoning behind this and I am fine with it. But I also try to avoid having things be async if they don't have to be. The safe bet is to probably keep it async and then we can pull that back before 1.0 if it truly isn't being used as async

kflansburg replied:
Sounds good.

kflansburg commented:
@thomastaylor312 @itowlson  I've added the wascc provider and cleaned up the outstanding panics to push this across the finish line. I think there are some opportunities to improve here by:

1. Refactoring some of the larger state handlers into multiple methods / macros for error transitions.
1. Pod::Running and container::Terminated seem ripe for a generic implementation. 

I would prefer to merge for now though and open tickets to keep this contained (no pun intended). I would prefer to merge this before starting any work on pulling out an operator framework.

itowlson commented on crates/wascc-provider/src/states/container/waiting.rs:
```diff
@@ -0,0 +1,272 @@
+                    )
+                }
+            }
+        };
```
~I find this much nesting really hard to parse.  Can we simplify this in any way?~  Looking at the broader context I now suspect this is existing code and we are just moving it into a state with the minimum possible change and risk, so can probably ignore this for now.

kflansburg replied:
I definitely think these things should be broken up in the future.

itowlson commented on crates/wascc-provider/src/states/container/waiting.rs:
```diff
@@ -0,0 +1,272 @@
+        }
+
+        Transition::next(self, Running)
+    }
```
Unless I've missed something, this is a 160-line method with up to 7 levels of indentation in places. Maybe I am dumb but I find that really hard to keep straight in my mind.  I assume that this is probably code that already exists and we are just moving it into a state, so I can see that we might not want to change it as part of this PR, but it should be high on some list for refactoring!

kflansburg replied:
Yep, I essentially just pulled in `start_container` and included any necessary context from `Pod::starting`. I think these should be broken up in future refactoring.

itowlson commented on crates/kubelet/src/state/common/mod.rs:
```diff
@@ -50,21 +52,21 @@ pub trait GenericProviderState: 'static + Send + Sync {
     /// Stores the pod module binaries for future execution. Typically your
     /// implementation can just move the modules map into a member field.
-    fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
+    async fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
```
Making these async feels like it should be a separate design discussion and a separate commit, not rolled into the container state changes.

kflansburg replied:
I think the reason this triggered the discussion is that the container state machines need access to these data structures concurrently. I think there is an argument to be made for moving the image pull into the container state machine so that they do not need this shared reference. 

I think we should establish a policy of all-async, but we can discuss that later.

itowlson commented on crates/kubelet/src/state/common/mod.rs:
```diff
@@ -73,11 +75,11 @@ pub trait GenericProvider: 'static + Send + Sync {
     type ProviderState: GenericProviderState;
     /// The state that is passed between Pod state handlers.
-    type PodState: GenericPodState;
+    type PodState: GenericPodState + ResourceState<SharedState = Self::ProviderState>;
```
Again I kind of feel like this is changing the design of one component in a PR that is concerned with another component.  But I guess that it emerges from a need to model shared state for containers.

kflansburg replied:
This was a little weird because the shared state type does need to be specified in `ResourceState`, but it was also needed on `Provider` for ergonomics. This line was changed to enforce that both of those types were the same.

itowlson commented on crates/kubelet/src/kubelet.rs:
```diff
@@ -255,16 +256,18 @@ mod test {
+    impl ResourceState for PodState {
+        type Manifest = Pod;
+        type Status = Status;
+        type SharedState = ProviderState;
```
This is an elegant solution to the `async_drop` problem and I like it **much** better than mine - thank you!

itowlson reviewed (approved):
This is great to see.  Thanks!

kflansburg commented:
Thanks for reviewing @itowlson @thomastaylor312 !
